### PR TITLE
julia_18-bin: add darwin support

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

As in #181772.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
So why are the tests disabled?

**Review by NickCao — APPROVED** [review]
As the darwin builds involves (nearly) no patching and mangling, I think we can trust upstream to provide at least working binaries. Cutting the trouble of running the lengthy and likely failing tests. But if they somehow magically pass without us doing anything, I'd like to flip that switch on.

**Comment by wegank** [comment]
@ninjin I uploaded the test log [here](https://gist.github.com/wegank/79e4dc0a4ff7fa824c32fd6533a3bf8e). I don't see any major issue with numerical functions in 1.18, since all linear algebra tests pass. Also, the first error actually comes from `download`.

**Comment by NickCao** [comment]
Shouldn't network related tests be automatically disabled when there's no network access, just like the source builds?

**Comment by wegank** [comment]
I don't think this is the case. #181772 also included a patch for this.

File: pkgs/development/compilers/julia/1.8-bin.nix [hunk 1]
```diff
@@ -1,5 +1,28 @@
...
+  # https://github.com/JuliaLang/julia/blob/master/test/choosetests.jl
+  skip_tests = [
+    # Tests require network access
+    "Artifacts"
+    "Downloads"
+    "LazyArtifacts"
+    "LibCURL"
+    "LibGit2"
+    "Sockets"
+    "download"
```
Comment (samuela) [inline pkgs/development/compilers/julia/1.8-bin.nix:13]: IIUC these tests are currently being run on linux. why disable them universally now?

Comment (NickCao) [reply pkgs/development/compilers/julia/1.8-bin.nix:13]: They are automatically disabled on linux.

Comment (wegank) [reply pkgs/development/compilers/julia/1.8-bin.nix]: Fixed by backporting JuliaLang/Julia#47914.

File: pkgs/development/compilers/julia/1.8-bin.nix [hunk 2]
```diff
@@ -13,8 +28,21 @@ stdenv.mkDerivation rec {
...
+    aarch64-darwin = fetchurl {
+      url = "https://julialang-s3.julialang.org/bin/mac/aarch64/${lib.versions.majorMinor version}/julia-${version}-macaarch64.tar.gz";
+      sha256 = "sha256-6oXgSJw2MkxNpiFjqhuC/PL1L3LRc+590hOjqSmSyrc=";
+    };
   }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
 
+  patches = [
+    # https://github.com/JuliaLang/julia/commit/f5eeba35d9bf20de251bb9160cc935c71e8b19ba
+    ./patches/1.8-bin/0001-allow-skipping-internet-required-tests.patch
+  ];
```
Comment (NickCao) [inline pkgs/development/compilers/julia/1.8-bin.nix:44]: Or just use fetchpatch here, the vendored patches are mostly nixpkgs specific.

Comment (wegank) [reply pkgs/development/compilers/julia/1.8-bin.nix:44]: That commit is for Julia 1.10, and does not apply to 1.8.x...

**Comment by NickCao** [comment]
@ofborg build julia_18-bin

**Comment by NickCao** [comment]
The `channels` test is once again flaky on aarch64-linux, maybe disable it unconditionally (what we are already doing with the source build) while you are at it.

**Review by samuela — APPROVED** [review]

**Review by NickCao — APPROVED** [review]

**Pull request merged**